src/monitors/WhammMonitor.v3

```diff
@@ -66,46 +70,226 @@ class WhammMonitor(whamm: Module) extends Monitor {
+                def memarg = parser.readMemoryArg();
+                vals.put(Values.u_v(memarg.flags));
+                vals.put(Values.u_v(memarg.memory_index));
+                vals.put(Values.u_v(u32.!(memarg.offset)));
+            	}
+            	MEMORY_INDEX      => vals.put(Values.u_v(parser.readU32("memory index", u32.!(-1))));
+            	HEAP_TYPE         => vals.put(Values.i_v(parser.readTypeCode()));
+            	SIG_INDEX         => vals.put(Values.u_v(parser.readU32("sig index", u32.!(-1))));
+            	STRUCT_TYPE_INDEX => vals.put(Values.u_v(parser.readU32("struct type index", u32.!(-1))));
+            	TABLE_INDEX       => vals.put(Values.u_v(parser.readU32("table index", u32.!(-1))));
+            	VALUE_TYPES       => ;		// TODO
+            	ZERO_BYTE         => vals.put(Values.u_v(parser.decoder.read1()));
+            	U32               => vals.put(Values.u_v(parser.readU32("imm U32", u32.!(-1))));
+            	I32               => vals.put(Values.i_v(parser.readI32()));
+            	I64               => vals.put(Values.l_v(parser.readI64()));
+            	F32               => vals.put(Values.f_v(parser.readF32()));
+            	F64               => vals.put(Values.d_v(parser.readF64()));
+            	V128              => vals.put(Values.s_v(parser.readF64(), parser.readF64()));  // TODO incorrect?
+            	LANE_INDEX        => vals.put(Values.u_v(parser.decoder.read1()));
+            	BR_CAST           => ;    // TODO
+            	CATCHES           => ;		// TODO
+            	CONT_INDEX        => ;		// TODO
+            	EX_HANDLERS       => ;		// TODO
+            	SUS_HANDLERS      => ;		// TODO
+            }
+          }
+
+          if (j >= vals.length) {
+            match (op) {
+              BR_TABLE, TRY_TABLE => return null;
+              _ => err.at(cv.parser.decoder).WhammProbeError("whamm probe", "immediate argument out of bounds");
+            }
+          }
+
+          arg = WhammArg.Val(vals[j]);
+          var t: ValueType;
+          match (vals[j]) {
+            Ref(_)     => ; // TODO
+            I31(_)     => ; // TODO
+            I32(_)     => t = ValueType.I32;
+            I64(_)     => t = ValueType.I64;
+            F32(_)     => t = ValueType.F32;
+            F64(_)     => t = ValueType.F64;
+            V128(_, _) => t = ValueType.V128;
+          }
+          if (expParams[i] != t) {
+            err.at(cv.parser.decoder).WhammProbeError("whamm probe", "type mismatch on immediate return type");
+          }
+				}
+				Arg(orig, i)   => {  // arg0 = first arg to bytecode
+				  def io = Opcodes.stackDelta(cv.sig, op, dupParser(cv));
+				  def args = io.0;
+				  if (i >= args) {
+				    match (op) {
+              CALL, CALL_INDIRECT => return null; // no match
+              _ => err.at(cv.parser.decoder).WhammProbeError("whamm probe", "instruction argument out of bounds");
+				    }
+				  }
+			    def stk = cv.val_stack.elems;
```

Comment:
I think this will be a little more robust to do a `.peekn` and use the parameter arity from `stackDelta`.